src/datadog/config_manager.cpp

```diff
@@ -1,80 +1,164 @@
   std::lock_guard<std::mutex> lock(mutex_);
-  return current_trace_sampler_;
+  return trace_sampler_;
 }
 
 std::shared_ptr<const SpanDefaults> ConfigManager::span_defaults() {
   std::lock_guard<std::mutex> lock(mutex_);
-  return current_span_defaults_;
+  return span_defaults_;
 }
 
 bool ConfigManager::report_traces() {
   std::lock_guard<std::mutex> lock(mutex_);
-  return current_report_traces_;
+  return report_traces_;
 }
 
-void ConfigManager::update(const ConfigUpdate& conf) {
+std::vector<ConfigMetadata> ConfigManager::update(const ConfigUpdate& conf) {
+  std::vector<ConfigMetadata> telemetry_conf;
+
   std::lock_guard<std::mutex> lock(mutex_);
 
-  if (conf.trace_sampler) {
-    if (auto finalized_trace_sampler_cfg =
-            finalize_config(*conf.trace_sampler)) {
-      current_trace_sampler_ =
-          std::make_shared<TraceSampler>(*finalized_trace_sampler_cfg, clock_);
-    } else {
-      // TODO: report error
+  if (conf.trace_sampling_rate) {
+    ConfigMetadata trace_sampling_metadata(
+        ConfigName::TRACE_SAMPLING_RATE, to_string(*conf.trace_sampling_rate),
+        ConfigMetadata::Origin::REMOTE_CONFIG);
+
+    TraceSamplerConfig trace_sampler_cfg;
+    trace_sampler_cfg.sample_rate = *conf.trace_sampling_rate;
+
+    auto finalized_trace_sampler_cfg = finalize_config(trace_sampler_cfg);
+    if (auto error = finalized_trace_sampler_cfg.if_error()) {
+      trace_sampling_metadata.error = *error;
+    }
+
+    auto trace_sampler =
+        std::make_shared<TraceSampler>(*finalized_trace_sampler_cfg, clock_);
+
+    if (trace_sampler != trace_sampler_.get()) {
+      trace_sampler_ = std::move(trace_sampler);
+      telemetry_conf.emplace_back(std::move(trace_sampling_metadata));
     }
   } else {
-    current_trace_sampler_ = default_trace_sampler_;
+    if (!trace_sampler_.is_default()) {
+      trace_sampler_.reset();
+      telemetry_conf.emplace_back(
+          default_config_metadata_[ConfigName::TRACE_SAMPLING_RATE]);
+    }
   }
 
   if (conf.tags) {
-    auto new_span_defaults =
-        std::make_shared<SpanDefaults>(*current_span_defaults_);
-    new_span_defaults->tags = std::move(*conf.tags);
+    ConfigMetadata tags_metadata(ConfigName::TAGS, join(*conf.tags, ","),
+                                 ConfigMetadata::Origin::REMOTE_CONFIG);
 
-    current_span_defaults_ = new_span_defaults;
+    auto parsed_tags = parse_tags(*conf.tags);
+    if (auto error = parsed_tags.if_error()) {
+      tags_metadata.error = *error;
+    }
+
+    if (*parsed_tags != span_defaults_.get()->tags) {
+      auto new_span_defaults =
+          std::make_shared<SpanDefaults>(*span_defaults_.get());
+      new_span_defaults->tags = std::move(*parsed_tags);
+
+      span_defaults_ = new_span_defaults;
+      telemetry_conf.emplace_back(std::move(tags_metadata));
+    }
   } else {
-    current_span_defaults_ = default_span_defaults_;
+    if (!span_defaults_.is_default()) {
```

Comment:
`else if`?

Reply:
addressed in e2bb8f465b2245055cb150d6d5bc8bcf69211c65